src/core/graph.go

```diff
@@ -184,6 +216,66 @@ func (graph *BuildGraph) linkDependencies(fromTarget, toTarget *BuildTarget) {
 	}
 }
 
+// checkArchitectureCompatibility checks that two targets have compatible architectures. It panics if they don't.
+func checkArchitectureCompatibility(from, to *BuildTarget) {
+	if from.Label.Arch != "" && to.Label.Arch != "" && from.Label.Arch != to.Label.Arch {
+		panic(fmt.Sprintf("%s requires a dependency on %s, but it's not available for that architecture", from.Label, to.Label.toArch("")))
+	}
+}
+
+// cloneTargetForArch returns a build target for the given architecture. It's otherwise
+// identical to the original one.
+func (graph *BuildGraph) cloneTargetForArch(target *BuildTarget, arch string) *BuildTarget {
+	archLabel := target.Label.toArch(arch)
+	t, present := graph.targets[archLabel]
+	if present {
+		return t
+	}
+	graph.targets[archLabel] = target // Sentinel, so we don't recurse into here incorrectly.
```

Comment:
There is, `graph.mutex` is already acquired at this point, and Go mutexes aren't re-entrant so we can't lock again. 
Go doesn't have any official or unofficial way of annotating that as a requirement though that I know of - Google used to have some stuff internally for C++ that was similar to [Clang's](http://clang.llvm.org/docs/ThreadSafetyAnalysis.html), it'd be sweet if they had something similar.

I'll add a comment to that effect, anyway.